--- a/blob/s3blob/s3blob.go	2026-04-09 22:48:10.700372961 +0000
+++ b/blob/s3blob/s3blob.go	2026-04-09 22:48:19.654602364 +0000
@@ -399,7 +399,7 @@
 	if b.useV2 {
 		in := &s3v2.ListObjectsV2Input{
 			Bucket:  aws.String(b.name),
-			MaxKeys: int32(pageSize),
+			MaxKeys: aws.Int32(int32(pageSize)),
 		}
 		if len(opts.PageToken) > 0 {
 			in.ContinuationToken = aws.String(string(opts.PageToken))
@@ -425,7 +425,7 @@
 				page.Objects[i] = &driver.ListObject{
 					Key:     unescapeKey(aws.StringValue(obj.Key)),
 					ModTime: *obj.LastModified,
-					Size:    obj.Size,
+					Size:    aws.Int64Value(obj.Size),
 					MD5:     eTagToMD5(obj.ETag),
 					AsFunc: func(i interface{}) bool {
 						p, ok := i.(*typesv2.Object)
@@ -576,7 +576,7 @@
 	var nextContinuationToken *string
 	if legacyResp.NextMarker != nil {
 		nextContinuationToken = legacyResp.NextMarker
-	} else if legacyResp.IsTruncated {
+	} else if aws.BoolValue(legacyResp.IsTruncated) {
 		nextContinuationToken = aws.String(aws.StringValue(legacyResp.Contents[len(legacyResp.Contents)-1].Key))
 	}
 	return &s3v2.ListObjectsV2Output{
@@ -706,7 +706,7 @@
 			Metadata:           md,
 			// CreateTime not supported; left as the zero time.
 			ModTime: aws.TimeValue(resp.LastModified),
-			Size:    resp.ContentLength,
+			Size:    aws.Int64Value(resp.ContentLength),
 			MD5:     eTagToMD5(resp.ETag),
 			ETag:    aws.StringValue(resp.ETag),
 			AsFunc: func(i interface{}) bool {
@@ -803,7 +803,7 @@
 			attrs: driver.ReaderAttributes{
 				ContentType: aws.StringValue(resp.ContentType),
 				ModTime:     aws.TimeValue(resp.LastModified),
-				Size:        getSize(resp.ContentLength, aws.StringValue(resp.ContentRange)),
+				Size:        getSize(aws.Int64Value(resp.ContentLength), aws.StringValue(resp.ContentRange)),
 			},
 			rawV2: resp,
 		}, nil
