From 6ce6a644e561f22021d031bc99f7b54a8d1142ad Mon Sep 17 00:00:00 2001
From: Oleksandr Ivantsiv <oivantsiv@nvidia.com>
Date: Mon, 10 Oct 2022 15:46:31 +0300
Subject: [PATCH] Allow warning during package compilation.

---
 debian/rules | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/debian/rules b/debian/rules
index 01e76c4..f4460af 100755
--- a/debian/rules
+++ b/debian/rules
@@ -8,13 +8,13 @@ export DEB_BUILD_OPTIONS=nostrip
 	dh $@  
 
 override_dh_auto_build:
-	./build.sh
+	./build.sh --allow_warnings
 	./scripts/cp_pkg_content.sh deb ./debian/tmp
 
 clean:
 	rm -f ./debian/debhelper-build-stamp ./debian/files ./debian/flexio.debhelper.log
 	rm -rf ./debian/flexio.substvars ./debian/tmp ./debian/flexio
-	./build.sh --clean
+	./build.sh --clean --allow_warnings
 
 override_dh_prep:
 	@ echo DH PREP
-- 
2.25.1

