From d2168682c00d3738d458a4f2bce23595e0e0729c Mon Sep 17 00:00:00 2001
From: Oleksandr Ivantsiv <oivantsiv@nvidia.com>
Date: Fri, 18 Nov 2022 17:29:38 +0200
Subject: [PATCH] Remove meson from dependencies.

---
 debian/control | 1 -
 1 file changed, 1 deletion(-)

diff --git a/debian/control b/debian/control
index b8a9ca1..d2367e3 100644
--- a/debian/control
+++ b/debian/control
@@ -6,7 +6,6 @@ Build-Depends: autotools-dev,
 	debhelper (>= 10), 
 	clang, 
 	lld, 
-	meson, 
 	patchelf
 #Standards-Version: 4.1.2
 Homepage: www.nvidia.com
-- 
2.25.1

