From 191df972f207774f8703ebf7fffea1129c11ca68 Mon Sep 17 00:00:00 2001
From: ab952469 <asha.behera@broadcom.com>
Date: Thu, 13 Mar 2025 14:46:03 +0530
Subject: [PATCH] Export remote info for authorization. authorization.

---
 auth.c         |   14 ++++++++++++++
 auth.h         |    3 +++
 session.c      |    3 +++
 sshd-session.c |    5 +++++
 4 files changed, 25 insertions(+)

diff --git a/auth.c b/auth.c
index 4b87886..1519c9a 100644
--- a/auth.c
+++ b/auth.c
@@ -779,3 +779,17 @@ auth_restrict_session(struct ssh *ssh)
 		fatal_f("failed to restrict session");
 	sshauthopt_free(restricted);
 }
+
+/* Export remote IP address and port for authorization. */
+	void
+export_remote_info(struct ssh *ssh)
+{
+	if (ssh != NULL)
+	{
+		const char *remote_ip = ssh_remote_ipaddr(ssh);
+		const int remote_port = ssh_remote_port(ssh);
+		const char remote_addr_port[32 + INET6_ADDRSTRLEN];
+		snprintf(remote_addr_port, sizeof(remote_addr_port), "%s %d", remote_ip, remote_port);
+		setenv("SSH_CLIENT_IPADDR_PORT", remote_addr_port, 1);
+	}
+}
diff --git a/auth.h b/auth.h
index 59799a8..0ad8eb0 100644
--- a/auth.h
+++ b/auth.h
@@ -242,6 +242,9 @@ FILE	*auth_openprincipals(const char *, struct passwd *, int);
 
 int	 sys_auth_passwd(struct ssh *, const char *);
 
+/* Export remote IP address and port for authorization. */
+void	 export_remote_info(struct ssh *);
+
 #if defined(KRB5) && !defined(HEIMDAL)
 krb5_error_code ssh_krb5_cc_gen(krb5_context, krb5_ccache *);
 #endif
diff --git a/session.c b/session.c
index 73e9b4a..1c822a4 100644
--- a/session.c
+++ b/session.c
@@ -629,6 +629,9 @@ do_exec_pty(struct ssh *ssh, Session *s, const char *command)
 		/* Close the extra descriptor for the pseudo tty. */
 		close(ttyfd);
 
+		/* Export remote IP address and port for authorization. */
+		export_remote_info(ssh);
+
 		/* record login, etc. similar to login(1) */
 #ifndef HAVE_OSF_SIA
 		do_login(ssh, s, command);
diff --git a/sshd-session.c b/sshd-session.c
index 2b6d2a9..03b119b 100644
--- a/sshd-session.c
+++ b/sshd-session.c
@@ -966,6 +966,8 @@ main(int ac, char **av)
 			fatal("test/dump modes not supported");
 			break;
 		case 'C':
+			/* Export remote IP address and port for authorization. */
+			export_remote_info(ssh);
 			connection_info = server_get_connection_info(ssh, 0, 0);
 			if (parse_server_match_testspec(connection_info,
 			    optarg) == -1)
@@ -1241,6 +1243,9 @@ main(int ac, char **av)
 	 */
 	remote_ip = ssh_remote_ipaddr(ssh);
 
+	/* Export remote IP address and port for authorization. */
+	export_remote_info(ssh);
+
 #ifdef SSH_AUDIT_EVENTS
 	audit_connection_from(remote_ip, remote_port);
 #endif
