From a05f213343ee7ee5dbfcfd1984c40db5c262db3c Mon Sep 17 00:00:00 2001
From: stormliang <stormliang@microsoft.com>
Date: Mon, 19 Jun 2023 13:57:01 +0000
Subject: [PATCH] From c423bce4db804c1d07d65ce3d06a9e62c4eceb2b Mon Sep 17
 00:00:00 2001 Subject: [PATCH] change log level for graceful restart events
diff --git a/bgpd/bgp_fsm.c b/bgpd/bgp_fsm.c
index 504343994..f58ab7c02 100644
--- a/bgpd/bgp_fsm.c
+++ b/bgpd/bgp_fsm.c
@@ -739,9 +739,9 @@ static void bgp_graceful_restart_timer_expire(struct event *thread)
 	afi_t afi;
 	safi_t safi;

-	if (bgp_debug_neighbor_events(peer)) {
-		zlog_debug("%pBP graceful restart timer expired", peer);
-		zlog_debug("%pBP graceful restart stalepath timer stopped",
+	if (peer) {
+		zlog_info("%pBP graceful restart timer expired", peer);
+		zlog_info("%pBP graceful restart stalepath timer stopped",
 			   peer);
 	}

@@ -801,8 +801,8 @@ static void bgp_graceful_stale_timer_expire(struct event *thread)
 	afi_t afi;
 	safi_t safi;

-	if (bgp_debug_neighbor_events(peer))
-		zlog_debug("%pBP graceful restart stalepath timer expired",
+	if (peer)
+		zlog_info("%pBP graceful restart stalepath timer expired",
 			   peer);

 	/* NSF delete stale route */
@@ -1380,20 +1380,17 @@ enum bgp_fsm_state_progress bgp_stop(struct peer_connection *connection)
 		/* graceful restart */
 		if (connection->t_gr_stale) {
 			EVENT_OFF(connection->t_gr_stale);
-			if (bgp_debug_neighbor_events(peer))
-				zlog_debug(
+				zlog_info(
 					"%pBP graceful restart stalepath timer stopped",
 					peer);
 		}
 		if (CHECK_FLAG(peer->sflags, PEER_STATUS_NSF_WAIT)) {
-			if (bgp_debug_neighbor_events(peer)) {
-				zlog_debug(
-					"%pBP graceful restart timer started for %d sec",
-					peer, peer->v_gr_restart);
-				zlog_debug(
-					"%pBP graceful restart stalepath timer started for %d sec",
-					peer, peer->bgp->stalepath_time);
-			}
+			zlog_info(
+				"%pBP graceful restart timer started for %d sec",
+				peer, peer->v_gr_restart);
+			zlog_info(
+				"%pBP graceful restart stalepath timer started for %d sec",
+				peer, peer->bgp->stalepath_time);
 			BGP_TIMER_ON(connection->t_gr_restart,
 				     bgp_graceful_restart_timer_expire,
 				     peer->v_gr_restart);
@@ -2216,8 +2213,7 @@ bgp_establish(struct peer_connection *connection)
 		UNSET_FLAG(peer->sflags, PEER_STATUS_NSF_MODE);
 		if (connection->t_gr_stale) {
 			EVENT_OFF(connection->t_gr_stale);
-			if (bgp_debug_neighbor_events(peer))
-				zlog_debug(
+			zlog_info(
 					"%pBP graceful restart stalepath timer stopped",
 					peer);
 		}
@@ -2225,8 +2221,7 @@ bgp_establish(struct peer_connection *connection)

 	if (connection->t_gr_restart) {
 		EVENT_OFF(connection->t_gr_restart);
-		if (bgp_debug_neighbor_events(peer))
-			zlog_debug("%pBP graceful restart timer stopped", peer);
+		zlog_info("%pBP graceful restart timer stopped", peer);
 	}

 	/* Reset uptime, turn on keepalives, send current table. */
diff --git a/bgpd/bgpd.c b/bgpd/bgpd.c
index d6d874be2..337879a2d 100644
--- a/bgpd/bgpd.c
+++ b/bgpd/bgpd.c
@@ -2629,13 +2629,11 @@ void peer_nsf_stop(struct peer *peer)

 	if (peer->connection->t_gr_restart) {
 		EVENT_OFF(peer->connection->t_gr_restart);
-		if (bgp_debug_neighbor_events(peer))
-			zlog_debug("%pBP graceful restart timer stopped", peer);
+		zlog_info("%pBP graceful restart timer stopped", peer);
 	}
 	if (peer->connection->t_gr_stale) {
 		EVENT_OFF(peer->connection->t_gr_stale);
-		if (bgp_debug_neighbor_events(peer))
-			zlog_debug(
+		zlog_info(
 				"%pBP graceful restart stalepath timer stopped",
 				peer);
 	}
