From 501f267ec4c7359fc518ba31c5d63045c935045c Mon Sep 17 00:00:00 2001
From: dgsudharsan <sudharsand@nvidia.com>
Date: Wed, 4 Dec 2024 20:39:46 +0000
Subject: [PATCH] Set multipath to 514 and disable bgp-vnc for optimization


diff --git a/debian/rules b/debian/rules
index e0b626fbf..41b6c9f57 100755
--- a/debian/rules
+++ b/debian/rules
@@ -73,8 +73,7 @@ override_dh_auto_configure:
 		--disable-protobuf \
 		--disable-zeromq \
 		--enable-ospfapi \
-		--enable-bgp-vnc \
-		--enable-multipath=256 \
+		--enable-multipath=514 \
 		--enable-pcre2posix \
 		\
 		--enable-user=frr \
-- 
2.43.2

