From 59d9d28b95ac5fae9089339758676764c58c9a70 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Mon, 14 Apr 2025 18:49:51 -0500
Subject: [PATCH] zebra: Translate tableid for dplane route notify

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 zebra/zebra_rib.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/zebra/zebra_rib.c b/zebra/zebra_rib.c
index af9e6b717..699156c4c 100644
--- a/zebra/zebra_rib.c
+++ b/zebra/zebra_rib.c
@@ -4881,7 +4881,7 @@ static inline void zebra_rib_translate_ctx_from_dplane(struct zebra_dplane_ctx *
 	if (vrfid == VRF_UNKNOWN)
 		dplane_ctx_set_vrf(ctx, zebra_vrf_lookup_by_table(tableid, nsid));
 	else if ((op == DPLANE_OP_ROUTE_INSTALL || op == DPLANE_OP_ROUTE_UPDATE ||
-		  op == DPLANE_OP_ROUTE_DELETE) &&
+		  op == DPLANE_OP_ROUTE_DELETE || op == DPLANE_OP_ROUTE_NOTIFY) &&
 		 tableid == ZEBRA_ROUTE_TABLE_UNKNOWN)
 		dplane_ctx_set_table(ctx, zebra_vrf_lookup_tableid(vrfid, nsid));
 }
-- 
2.34.1

