From b6caf88bd3bce9673d49435453991c49712287aa Mon Sep 17 00:00:00 2001
From: Rajasekar Raja <rajasekarr@nvidia.com>
Date: Thu, 11 Apr 2024 22:27:37 -0700
Subject: [PATCH 08/11] zebra: backpressure - Fix Null ptr access (Coverity
 Issue)

Fix dereferencing NULL ptr making coverity happy.

Ticket :#3390099

Signed-off-by: Rajasekar Raja <rajasekarr@nvidia.com>

diff --git a/bgpd/bgp_zebra.c b/bgpd/bgp_zebra.c
index b81acaf8ec..524551b1e0 100644
--- a/bgpd/bgp_zebra.c
+++ b/bgpd/bgp_zebra.c
@@ -1809,8 +1809,7 @@ static void bgp_handle_route_announcements_to_zebra(struct thread *e)
 		table = bgp_dest_table(dest);
 		install =
 			CHECK_FLAG(dest->flags, BGP_NODE_SCHEDULE_FOR_INSTALL);
-		if (table && table->afi == AFI_L2VPN &&
-		    table->safi == SAFI_EVPN)
+		if (table->afi == AFI_L2VPN && table->safi == SAFI_EVPN)
 			is_evpn = true;
 
 		if (BGP_DEBUG(zebra, ZEBRA))
-- 
2.17.1

