diff --git a/debian/control b/debian/control
index 0ef86f4..3e8dfb1 100644
--- a/debian/control
+++ b/debian/control
@@ -65,7 +65,7 @@ Description: InfiniBand Communication Manager Assistant (ACM)
 Package: ibverbs-providers
 Architecture: linux-any
 Multi-Arch: same
-Depends: ${misc:Depends}, ${shlibs:Depends}
+#Depends: ${misc:Depends}, ${shlibs:Depends}
 Provides: libefa1, libipathverbs1, libmana1, libmlx5-1, libmthca1
 Replaces: libipathverbs1,
           libmlx5-1,
@@ -138,7 +138,7 @@ Section: libs
 Pre-Depends: ${misc:Pre-Depends}
 Depends: adduser, ${misc:Depends}, ${shlibs:Depends}
 Recommends: ibverbs-providers
-Breaks: ibverbs-providers (<< 34~)
+#Breaks: ibverbs-providers (<< 34~)
 Description: Library for direct userspace use of RDMA (InfiniBand/iWARP)
  libibverbs is a library that allows userspace processes to use RDMA
  "verbs" as described in the InfiniBand Architecture Specification and
