From 719292c4800054ead278affc2972d8f77babd227 Mon Sep 17 00:00:00 2001
From: rajib-dutta1 <rajib.dutta@dell.com>
Date: Wed, 10 Jan 2024 20:51:03 -0800
Subject: [PATCH] Changes needed to install enterprise.txt

---
 Makefile.am  | 40 ++++++++++++++++++++--------------------
 configure.ac | 30 +++++++++++++++---------------
 2 files changed, 35 insertions(+), 35 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 355d3f0..ec9605c 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -49,25 +49,25 @@ dist-hook:
 .PHONY: install-pen-database uninstall-pen-database
 .INTERMEDIATE: %.o %.la enterprise-numbers
 
-#if DOWNLOAD
-
-#enterprise-numbers:
-#	@echo Downloading IANA PEN database...
-#	@$(DOWNLOAD) "$(IANA_PEN)" > tmpfile.$$PPID || {\
-#		echo "FAILED to download the IANA PEN database"; \
-#		rm tmpfile.$$PPID; \
-#		false; \
-#	}
-#	@mv tmpfile.$$PPID $@
-#
-#install-pen-database: enterprise-numbers
-#	mkdir -m 755 -p $(DESTDIR)$(IANADIR)
-#	$(INSTALL_DATA) $< $(DESTDIR)$(IANADIR)/
-#
-#uninstall-pen-database:
-#	-rm -rf $(DESTDIR)$(IANADIR)/enterprise-numbers
-#
-#else
+if DOWNLOAD
+
+enterprise-numbers.txt:
+	@echo Downloading IANA PEN database...
+	@$(DOWNLOAD) "$(IANA_PEN)" > tmpfile.$$PPID || {\
+		echo "FAILED to download the IANA PEN database"; \
+		rm tmpfile.$$PPID; \
+		false; \
+	}
+	@mv tmpfile.$$PPID $@
+
+install-pen-database: enterprise-numbers.txt
+	mkdir -m 755 -p $(DESTDIR)$(IANADIR)
+	$(INSTALL_DATA) $< $(DESTDIR)$(IANADIR)/
+
+uninstall-pen-database:
+	-rm -rf $(DESTDIR)$(IANADIR)/enterprise-numbers.txt
+
+else
 
 install-pen-database:
 	@echo "*** NOT installing the IANA PEN database."
@@ -77,7 +77,7 @@ uninstall-pen-database:
 	@echo "*** NOT uninstalling the IANA PEN database."
 	@echo "*** It was installed manually (if ever)."
 
-#endif
+endif
 
 install-data-local: install-pen-database
 	mkdir -p $(DESTDIR)$(DOCDIR)
diff --git a/configure.ac b/configure.ac
index b4321c6..d6ba62b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -56,22 +56,22 @@ if test "x$exec_prefix" = "xNONE"; then
 	exec_prefix="$prefix"
 fi
 
-#if test "x$WGET" = "x"; then
-#	if test "x$CURL" = "x"; then
-#		AC_MSG_WARN([** Neither wget nor curl could be found.])
-#		AC_MSG_WARN([** IANA PEN database will not be installed by `make install` !])
-#	else
-#		DOWNLOAD="$CURL --location --progress-bar"
-#		AM_CONDITIONAL([DOWNLOAD], [true])
-#	fi
-#else
-#	DOWNLOAD="$WGET -c -nd -O -"
-#	AM_CONDITIONAL([DOWNLOAD], [true])
-#fi
+if test "x$WGET" = "x"; then
+	if test "x$CURL" = "x"; then
+		AC_MSG_WARN([** Neither wget nor curl could be found.])
+		AC_MSG_WARN([** IANA PEN database will not be installed by `make install` !])
+	else
+		DOWNLOAD="$CURL --location --progress-bar"
+		AM_CONDITIONAL([DOWNLOAD], [true])
+	fi
+else
+	DOWNLOAD="$WGET -c -nd -O -"
+	AM_CONDITIONAL([DOWNLOAD], [true])
+fi
 
-#AC_MSG_WARN([** Download is:])
-#AC_MSG_WARN($DOWNLOAD)
-#AC_SUBST(DOWNLOAD, $DOWNLOAD)
+AC_MSG_WARN([** Download is:])
+AC_MSG_WARN($DOWNLOAD)
+AC_SUBST(DOWNLOAD, $DOWNLOAD)
 
 dnl
 dnl set default option values
-- 
2.25.1

