diff --git a/src/rshim_net.c b/src/rshim_net.c
index a9e8663..7214df2 100644
--- a/src/rshim_net.c
+++ b/src/rshim_net.c
@@ -263,7 +263,7 @@ int rshim_net_init(rshim_backend_t *bd)
   char ifname[IFNAMSIZ];
   int rc, fd[2];
 
-  snprintf(ifname, sizeof(ifname), "tmfifo_net%d", bd->index);
+  snprintf(ifname, sizeof(ifname), "dpu%d", bd->index);
   bd->net_fd = rshim_if_open(ifname, bd->index);
 
   if (bd->net_fd < 0)
