diff --git a/configs/profiles/connectx/debian/control b/configs/profiles/connectx/debian/control
index 62f9403..fea0155 100644
--- a/configs/profiles/connectx/debian/control
+++ b/configs/profiles/connectx/debian/control
@@ -4,7 +4,6 @@ Maintainer: <networking-support@nvidia.com>
 Build-Depends: debhelper (>= 10~),
 	       gcc,
 	       g++,
-	       meson,
 	       pkg-config,
 	       libibverbs-dev,
 	       mlnx-dpdk-dev,
@@ -55,7 +54,6 @@ Architecture: amd64 arm64 armhf i386 ppc64el
 Depends: ${misc:Depends},
 	 libdoca-cx-libs-dev,
 	 gcc,
-	 meson,
 	 pkg-config,
 	 mlnx-dpdk-dev,
 	 libjson-c-dev,
@@ -71,7 +69,6 @@ Architecture: amd64 arm64 armhf i386 ppc64el
 Depends: ${misc:Depends},
 	 libdoca-cx-libs-dev,
 	 gcc,
-	 meson,
 	 pkg-config,
 	 mlnx-dpdk-dev,
 	 libjson-c-dev
