diff --git a/meson.build b/meson.build
index 06e71ab..c674944 100644
--- a/meson.build
+++ b/meson.build
@@ -16,7 +16,7 @@ project('DOCA', 'C', 'CPP',
 	version: run_command(find_program('cat', 'more'), files('VERSION'), check: true).stdout().strip(),
 	default_options: [
 		'buildtype=debug',
-		'werror=true',
+		'werror=false',
 		'warning_level=2',
 		'b_staticpic=true',
 		'prefix=/opt/mellanox/doca',
